diff -up octave-4.0.0/libgnu/math.in.h.signbit octave-4.0.0/libgnu/math.in.h
--- octave-4.0.0/libgnu/math.in.h.signbit	2016-02-19 19:16:24.525856660 -0700
+++ octave-4.0.0/libgnu/math.in.h	2016-02-19 19:32:27.001262926 -0700
@@ -2205,7 +2205,7 @@ _GL_WARN_REAL_FLOATING_DECL (isnan);
 
 
 #if @GNULIB_SIGNBIT@
-# if @REPLACE_SIGNBIT_USING_GCC@
+# if @REPLACE_SIGNBIT_USING_GCC@ && !defined(__cplusplus)
 #  undef signbit
    /* GCC 4.0 and newer provides three built-ins for signbit.  */
 #  define signbit(x) \
