diff -r d1db86336a49 scripts/miscellaneous/tar.m
--- a/scripts/miscellaneous/tar.m	Thu Aug 11 02:41:36 2011 -0500
+++ b/scripts/miscellaneous/tar.m	Thu Aug 11 10:56:18 2011 -0600
@@ -42,7 +42,7 @@
     files = cellstr (files);
   endif
 
-  if (ischar (tarfile) && iscellstr (files) && ischar (root))
+  if !(ischar (tarfile) && iscellstr (files) && ischar (root))
     error ("tar: all arguments must be character strings");
   endif
 
