diff -up octave-3.8.0/libinterp/corefcn/oct.h.config octave-3.8.0/libinterp/corefcn/oct.h
--- octave-3.8.0/libinterp/corefcn/oct.h.config	2013-12-27 14:57:41.000000000 -0700
+++ octave-3.8.0/libinterp/corefcn/oct.h	2014-01-08 14:10:27.816397648 -0700
@@ -28,7 +28,7 @@ along with Octave; see the file COPYING.
 // config.h needs to be first because it includes #defines that can */
 // affect other header files.
 
-#include <config.h>
+#include "config.h"
 
 #include "Matrix.h"
 
