--- octave-2.9.4/src/DLD-FUNCTIONS/spchol.cc.orig	2005-11-11 10:44:05.000000000 -0700
+++ octave-2.9.4/src/DLD-FUNCTIONS/spchol.cc	2005-11-30 15:41:35.000000000 -0700
@@ -609,7 +609,11 @@
 	  for (octave_idx_type k = 0 ; k < n ; k++)
 	    {
 	      // get the kth row of L and store in the columns of L
+#ifdef IDX_TYPE_LONG
+	      cholmod_l_row_subtree (A1, A2, k, Parent, R, cm) ;
+#else
 	      cholmod_row_subtree (A1, A2, k, Parent, R, cm) ;
+#endif
 	      for (octave_idx_type p = 0 ; p < Rp [1] ; p++)
 		L.xridx (W [Ri [p]]++) = k ;
 
