Index: scripts/miscellaneous/mex.m
===================================================================
RCS file: /cvs/octave/scripts/miscellaneous/mex.m,v
retrieving revision 1.1
diff -u -r1.1 mex.m
--- scripts/miscellaneous/mex.m 26 Jul 2006 17:19:10 -0000      1.1
+++ scripts/miscellaneous/mex.m 28 Jul 2006 21:06:42 -0000
@@ -20,7 +20,7 @@
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} mex [options] file ...
 ## Compile source code written in C, C++, or Fortran, to a MEX file.
-## This is equivalent to @code {mkoctfile --mex [options] file}.
+## This is equivalent to @code{mkoctfile --mex [options] file}.
 ## @seealso{mkoctfile}
 ## @end deftypefn

