Index: liboctave/mx-inlines.cc
===================================================================
RCS file: /cvs/octave/liboctave/mx-inlines.cc,v
retrieving revision 1.49
retrieving revision 1.50
diff -u -r1.49 -r1.50
--- liboctave/mx-inlines.cc	22 Aug 2006 18:37:43 -0000	1.49
+++ liboctave/mx-inlines.cc	26 Aug 2006 03:02:39 -0000	1.50
@@ -383,6 +383,12 @@
         } \
     } \
  \
+  if (nd == 2 && dv(0) == 0 && dv(1) == 0) \
+    { \
+      retval.resize (dim_vector (1, 1), INIT_VAL); \
+      return retval; \
+    } \
+ \
   /* We need to find first non-singleton dim.  */ \
  \
   if (dim == -1) \
