diff -r -u octave-3.0.3.orig/liboctave/oct-md5.cc octave-3.0.3/liboctave/oct-md5.cc
--- octave-3.0.3.orig/liboctave/oct-md5.cc	2008-09-24 00:13:47.000000000 -0700
+++ octave-3.0.3/liboctave/oct-md5.cc	2009-02-23 19:30:37.000000000 -0700
@@ -20,13 +20,15 @@
 
 */
 
-#include <string>
-#include <vector>
-
 #ifdef HAVE_CONFIG_H
 #include "config.h"
 #endif
 
+#include <cstdio>
+
+#include <string>
+#include <vector>
+
 #include "lo-error.h"
 #include "oct-md5.h"
 #include "md5.h"
diff -r -u octave-3.0.3.orig/scripts/gethelp.cc octave-3.0.3/scripts/gethelp.cc
--- octave-3.0.3.orig/scripts/gethelp.cc	2008-09-24 00:13:48.000000000 -0700
+++ octave-3.0.3/scripts/gethelp.cc	2009-02-23 19:30:37.000000000 -0700
@@ -24,8 +24,10 @@
 #define __USE_STD_IOSTREAM
 #endif
 
-#include <string>
+#include <cstdio>
+
 #include <iostream>
+#include <string>
 
 #ifndef NPOS
 #define NPOS std::string::npos
--- a/liboctave/sparse-util.cc	Fri Oct 10 11:35:10 2008 +0200
+++ b/liboctave/sparse-util.cc	Sun Oct 12 10:30:40 2008 +0200
@@ -30,8 +30,15 @@ along with Octave; see the file COPYING.
 #include "lo-error.h"
 #include "sparse-util.h"
 
+// FIXME this overload is here due to API change in SuiteSparse (3.1 -> 3.2)
 void
 SparseCholError (int status, char *file, int line, char *message)
+{
+  SparseCholError (status, file, line, message);
+}
+
+void
+SparseCholError (int status, const char *file, int line, const char *message)
 {
   (*current_liboctave_warning_handler)("warning %i, at line %i in file %s",
 				     status, line, file);
--- a/liboctave/sparse-util.h	Fri Oct 10 11:35:10 2008 +0200
+++ b/liboctave/sparse-util.h	Sun Oct 12 10:30:40 2008 +0200
@@ -24,7 +24,11 @@ along with Octave; see the file COPYING.
 #if !defined (octave_sparse_util_h)
 #define octave_sparse_util_h 1
 
-extern OCTAVE_API void SparseCholError (int status, char *file, int line, char *message);
+// FIXME this overload is here due to API change in SuiteSparse (3.1 -> 3.2)
+extern OCTAVE_API void SparseCholError (int status, char *file, 
+                                        int line, char *message);
+extern OCTAVE_API void SparseCholError (int status, const char *file, 
+                                        int line, const char *message);
 extern OCTAVE_API int SparseCholPrint (const char *fmt, ...);
 
 #endif

