diff -u octave-3.0.2.org/configure octave-3.0.2/configure
--- octave-3.0.2.org/configure	2008-10-23 20:03:19.000000000 +0530
+++ octave-3.0.2/configure	2008-10-23 20:03:44.000000000 +0530
@@ -11078,7 +11078,7 @@
 
           ;;
           # Ignore these flags.
-        -lang* | -lcrt*.o | -lc | -lgcc* | -lSystem | -libmil | -LANG:=* | -LIST:* | -LNO:*)
+        -little| -lang* | -lcrt*.o | -lc | -lgcc* | -lSystem | -libmil | -LANG:=* | -LIST:* | -LNO:*)
           ;;
         -lkernel32)
           test x"$CYGWIN" != xyes && ac_cv_f77_libs="$ac_cv_f77_libs $ac_arg"
